arches/app/src/components/ControlledListManager/ValueEditor.vue

```diff
@@ -0,0 +1,372 @@
+    }
+    if (!valueType) {
+        if (valueCategory === NOTE) {
+            // Show everything but labels for now.
+            // We're not returning category from the API, and images already excluded.
+            return item.value.values.filter(
+                value => ![PREF_LABEL, ALT_LABEL].includes(value.valuetype_id)
+            );
+        }
+    }
+    return item.value.values.filter(
+        value => value.valuetype_id === valueType
+    );
+});
+
+const onSave = async (event: DataTableRowEditInitEvent) => {
+    // normalize new value numbers (starting at 1000) to null
+    const normalizedNewData: Value = {
+        ...event.newData,
+        id: typeof event.newData.id === 'string' ? event.newData.id : null,
+        value: event.newData.value.trim(),
+    };
+    const upsertedValue: Value = await upsertValue(
+        normalizedNewData,
+        toast,
+        $gettext,
+    );
+    if (upsertedValue) {
+        if (normalizedNewData.id) {
+            updateItemValue(upsertedValue);
+        } else {
+            appendItemValue(upsertedValue);
+            removeItemValue(event.newData);
+        }
+    } else if (normalizedNewData.id === null) {
+        removeItemValue(event.newData);
+    }
+};
+
+const onDelete = async (value: NewValue | Value) => {
+    if (typeof value.id === 'number') {
+        removeItemValue(value);
+        return;
+    }
+    const deleted = await deleteValue(value, toast, $gettext);
+    if (deleted) {
+        removeItemValue(value);
+    }
+};
+
+const appendItemValue = (newValue: Value) => { item.value.values.push(newValue); };
+
+const removeItemValue = (removedValue: Value | NewValue) => {
+    const toDelete = item.value.values.findIndex(valueFromItem => valueFromItem.id === removedValue.id);
+    item.value.values.splice(toDelete, 1);
+};
+
+const updateItemValue = (updatedValue: Value) => {
+    const toUpdate = item.value.values.find(valueFromItem => valueFromItem.id === updatedValue.id);
+    if (toUpdate) {
+        toUpdate.language_id = updatedValue.language_id;
+        toUpdate.value = updatedValue.value;
+        toUpdate.valuetype_id = updatedValue.valuetype_id;
+    }
+};
+
+const onEdit = (event: DataTableRowEditInitEvent) => {
+    rowIndexToFocus.value = event.index;
+};
+
+const makeValueEditable = (clickedValue: Value, index: number) => {
+    if (!editingRows.value.includes(clickedValue)) {
+        editingRows.value = [ ...editingRows.value, clickedValue ];
+    }
+    if (index === -1) {
+        // Coming from <AddValue>
+        rowIndexToFocus.value = Math.max(values.value.length - 1, 0);
+    } else {
+        rowIndexToFocus.value = index;
+    }
+};
+
+const inputSelector = computed(() => {
+    return `[data-p-index="${rowIndexToFocus.value}"]`;
+});
+
+const focusInput = () => {
+    // The editor (pencil) button immediately hogs focus with a setTimeout of 1,
+    // so we'll get in line behind it to set focus to the input.
+    setTimeout(() => {
```

Comment:
why does the editor hog focus with a setTimeout? Is it possible to refactor things so one isn't need here or in the other component?

Reply:
The editor I'm referring to here is the Primevue `<DataTable>` component, which puts keyboard focus on the pencil icon when adding a new row, which I think is a good behavior in general. It happens to do that with setTimeout, which is out of my control.